In [14]:
## Load necessary packages
import os
import pandas as pd
import glob
import numpy as np
import networkx as nx
# import matplotlib.pyplot as plt

## load the needed tqdm package
from tqdm import tqdm

# Enable tqdm for pandas apply
tqdm.pandas()

## Define the version number
version_number = "01_06_2026"
deployment_date = "2026-01-06"

In [16]:
## Load all helper functions
%run /Users/Weiqi0/ISB_working/Hadlock_lab/QI_ISB_Git_repo/TranslatorPharcogenomicsKG/Parser_helper_functions.ipynb

In [17]:
## Notice!! Please change the file path of following codes into your own
raw_files_path = '/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/GtoPdb/'
## Note! define download path in the very end of the notebook!!

In [18]:
## Check all node files being read
## Read all BigGIM node csv file in group 1

for f in os.listdir(raw_files_path):
    if f.endswith('.csv'):
        print(f)

approved_drug_primary_target_interactions.csv
ligands.csv
interactions.csv
GtP_to_HGNC_mapping.csv
targets_and_families.csv
approved_drug_detailed_interactions.csv
peptides.csv


### 1. interactions.csv file
* Note: has already been ingested in tier 0 graph

In [57]:
# ## Load the file, skipping the metadata line
# ## Read each individual csv files
# interactions_df = pd.read_csv(raw_files_path + 'interactions.csv', skiprows=1)

# interactions_df.head(10)

/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_41880/4279433548.py:3: DtypeWarning: Columns (6,8,9,10,11,39) have mixed types. Specify dtype option on import or set low_memory=False.
  interactions_df = pd.read_csv(raw_files_path + 'interactions.csv', skiprows=1)


Target  Target ID  \
0                                        12S-LOX     1387.0   
1                                       15-LOX-1     1388.0   
2                                       15-LOX-1     1388.0   
3                                       15-LOX-1     1388.0   
4                                       15-LOX-1     1388.0   
5                                       15-LOX-2     1389.0   
6                                       15-LOX-2     1389.0   
7           3-Mercaptopyruvate sulfurtransferase     1446.0   
8  3-phosphoinositide dependent protein kinase 1     1519.0   
9  3-phosphoinositide dependent protein kinase 1     1519.0   

  Target Subunit IDs Target Gene Symbol Target UniProt ID  \
0                NaN             ALOX12            P18054   
1                NaN             ALOX15            P16050   
2                NaN             ALOX15            P16050   
3                NaN             ALOX15            P16050   
4                NaN             ALOX15            P16050   
5                NaN            ALOX15B            O15296   
6                NaN            ALOX15B            O15296   
7                NaN               MPST            P25325   
8                NaN              PDPK1            O15530   
9                NaN              PDPK1            O15530   

  Target Ensembl Gene ID Target Ligand  Target Ligand ID  \
0        ENSG00000108839           NaN               NaN   
1        ENSG00000161905           NaN               NaN   
2        ENSG00000161905           NaN               NaN   
3        ENSG00000161905           NaN               NaN   
4        ENSG00000161905           NaN               NaN   
5        ENSG00000179593           NaN               NaN   
6        ENSG00000179593           NaN               NaN   
7        ENSG00000128309           NaN               NaN   
8        ENSG00000140992           NaN               NaN   
9        ENSG00000140992           NaN               NaN   

  Target Ligand Subunit IDs Target Ligand Gene Symbol  ... Affinity Low  \
0                       NaN                       NaN  ...          NaN   
1                       NaN                       NaN  ...          NaN   
2                       NaN                       NaN  ...          NaN   
3                       NaN                       NaN  ...          NaN   
4                       NaN                       NaN  ...          NaN   
5                       NaN                       NaN  ...          NaN   
6                       NaN                       NaN  ...          NaN   
7                       NaN                       NaN  ...          NaN   
8                       NaN                       NaN  ...          NaN   
9                       NaN                       NaN  ...          NaN   

  Original Affinity Units  Original Affinity Low nm  \
0                    IC50                       NaN   
1                      Kd                       NaN   
2                      Ki                       NaN   
3                    IC50                       NaN   
4                       -                       NaN   
5                    IC50                       NaN   
6                      Ki                       NaN   
7                    IC50                       NaN   
8                    IC50                       NaN   
9                    IC50                       NaN   

  Original Affinity Median nm  Original Affinity High nm  \
0                       340.0                        NaN   
1                      3900.0                        NaN   
2                        10.0                        NaN   
3                       200.0                        NaN   
4                         NaN                        NaN   
5                        51.0                        NaN   
6                      2500.0                        NaN   
7                      2700.0                        NaN   
8                        33.0              

In [6]:
# interactions_df.head(5)

In [7]:
# list(interactions_df)

In [8]:
# ## check all unique combinations between subject & object
# from itertools import combinations

# # Get all combinations of 3 column names
# col_combos = ['Ligand Type', 'Type']

# unique_combinations = interactions_df[col_combos].drop_duplicates()

# print(unique_combinations)

In [9]:
# ## check all unique combinations between subject & object
# from itertools import combinations

# # Get all combinations of 3 column names
# col_combos = ['Target Species']

# unique_combinations = interactions_df[col_combos].drop_duplicates()

# print(unique_combinations)

### 1. targets_and_families.csv file
*

In [42]:
# ## Load the file, skipping the metadata line
# ## Read each individual csv files
# targets_families_df = pd.read_csv(raw_files_path + 'targets_and_families.csv', skiprows=1)

# targets_families_df.head(5)

Type  Family id                    Family name  Target id  \
0  gpcr          1  5-Hydroxytryptamine receptors          1   
1  gpcr          1  5-Hydroxytryptamine receptors          2   
2  gpcr          1  5-Hydroxytryptamine receptors          3   
3  gpcr          1  5-Hydroxytryptamine receptors          4   
4  gpcr          1  5-Hydroxytryptamine receptors          5   

                  Target name Subunit id Subunit name Target systematic name  \
0  5-HT<sub>1A</sub> receptor        NaN          NaN                    NaN   
1  5-HT<sub>1B</sub> receptor        NaN          NaN                    NaN   
2  5-HT<sub>1D</sub> receptor        NaN          NaN                    NaN   
3  5-ht<sub>1e</sub> receptor        NaN          NaN                    NaN   
4  5-HT<sub>1F</sub> receptor        NaN          NaN                    NaN   

  Target abbreviated name                                           synonyms  \
0                     NaN  ADRBRL1|5-HT1A|ADRB2RL1|serotonin receptor 1A|...   
1                     NaN  5-HT1B|5-HT1DB|HTR1D2|5-HT1B serotonin recepto...   
2                     NaN  5-HT<sub>1D&alpha;</sub>|HTRL|5-HT1D|HT1DA|ser...   
3                     NaN  5-HT<sub>1E</sub><sub>&alpha;</sub>|5-HT1E|5-h...   
4                     NaN  5-HT<sub>1E&beta;</sub>|5-HT<sub>6</sub>|5-HT1...   

   ...    Rat Ensembl Gene     MGI id MGI symbol  \
0  ...  ENSRNOG00000010254  MGI:96273      Htr1a   
1  ...  ENSRNOG00000013042  MGI:96274      Htr1b   
2  ...  ENSRNOG00000012038  MGI:96276      Htr1d   
3  ...                 NaN        NaN        NaN   
4  ...  ENSRNOG00000000716  MGI:99842      Htr1f   

                                      MGI name Mouse genetic localisation  \
0  5-hydroxytryptamine (serotonin) receptor 1A                13 56.92 cM   
1  5-hydroxytryptamine (serotonin) receptor 1B                 9 44.61 cM   
2  5-hydroxytryptamine (serotonin) receptor 1D                 4 68.74 cM   
3                                          NaN                        NaN   
4  5-hydroxytryptamine (serotonin) receptor 1F                 16 37.1 cM   

  Mouse nucleotide RefSeq Mouse protein RefSeq Mouse SwissProt  \
0               NM_008308            NP_032334          Q64264   
1               NM_010482            NP_034612          P28334   
2               NM_008309            NP_032335          Q61224   
3                     NaN                  NaN             NaN   
4               NM_008310            NP_032336          Q02284   

  Mouse Entrez Gene  Mouse Ensembl Gene  
0             15550  ENSMUSG00000021721  
1             15551  ENSMUSG00000049511  
2             15552  ENSMUSG00000070687  
3               NaN                 NaN  
4             15557  ENSMUSG00000050783  

[5 rows x 37 columns]

In [10]:
# list(targets_families_df)

In [11]:
# ## check all unique combinations between subject & object
# from itertools import combinations

# # Get all combinations of 3 column names
# col_combos = ['Type']

# unique_combinations = targets_families_df[col_combos].drop_duplicates()

# print(unique_combinations)

## 3. ligands.csv
* 

In [45]:
# ## Load the file, skipping the metadata line
# ## Read each individual csv files
# ligands_df = pd.read_csv(raw_files_path + 'ligands.csv', skiprows=1)

# ligands_df.head(5)

Ligand ID                            Name Species               Type  \
0       6069  10,10-difluoro TXA<sub>2</sub>     NaN  Synthetic organic   
1      13409                   10-Me-Aplog-1     NaN  Synthetic organic   
2       1835  10-OBn-7&alpha;-F-gingkolide B     NaN    Natural product   
3       1838         10-OBn-epi-ginkgolide C     NaN    Natural product   
4       1836             10-OBn-ginkgolide B     NaN    Natural product   

  Approved Withdrawn Labelled Radioactive  PubChem SID  PubChem CID  ...  \
0      NaN       NaN      NaN         NaN  178102690.0   73755171.0  ...   
1      NaN       NaN      NaN         NaN  496703379.0   60155209.0  ...   
2      NaN       NaN      NaN         NaN  135652650.0   73755010.0  ...   
3      NaN       NaN      NaN         NaN  135652651.0   73755013.0  ...   
4      NaN       NaN      NaN         NaN  135652652.0   73755011.0  ...   

  Ligand Subunit Ensembl IDs  \
0                        NaN   
1                        NaN   
2                        NaN   
3                        NaN   
4                        NaN   

                                          IUPAC name  INN  \
0  (Z)-7-[(1S,3R,4R,5R)-7,7-difluoro-3-[(E,3S)-3-...  NaN   
1                                                NaN  NaN   
2                                                NaN  NaN   
3                                                NaN  NaN   
4                                                NaN  NaN   

                                            Synonyms  \
0                                                NaN   
1  compound 1 [PMID: 32053979]|compound 5 [PMID: ...   
2  10-OBn-7&alpha;-F-GB|compound 17 [PMID: 12570381]   
3                                      10-OBn-epi-GC   
4                                          10-OBn-GB   

                                              SMILES  \
0  CCCCC[C@@H](/C=C/[C@H]1C[C@@H]2O[C@H]([C@@H]1C...   
1  [H][C@@]12CCC(C)(C)[C@]3(C[C@]([H])(OC(=O)C[C@...   
2  O=C1OC2C3([C@H]1OCc1ccccc1)C([C@@H](C1C43[C@@]...   
3  O=C1OC2C3([C@H]1OCc1ccccc1)C([C@@H](C1C43[C@@]...   
4  O=C1OC2C3([C@H]1OCc1ccccc1)C(CC1C43[C@@](O2)(C...   

                      InChIKey  \
0  CIYZUODQUHAGKZ-SCVXEUJASA-N   
1  SWKUTOXEFLDOID-PYCRRGESSA-N   
2  LFVHCLPCEWRXME-CLNXZACASA-N   
3  YPOMAFCMSAKBRX-CLNXZACASA-N   
4  PHJOWQWLQBFEEM-ZTPSGWAGSA-N   

                                               InChI GtoImmuPdb GtoMPdb  \
0  InChI=1S/C21H32F2O4/c1-2-3-6-9-16(24)13-12-15-...        NaN     NaN   
1  InChI=1S/C28H40O8/c1-18-23(10-5-4-7-19-8-6-9-2...        NaN     NaN   
2  InChI=1S/C27H29FO10/c1-11-19(30)35-17-15(29)25...        NaN     NaN   
3  InChI=1S/C27H30O11/c1-11-19(30)35-17-15(29)25-...        NaN     NaN   
4  InChI=1S/C27H30O10/c1-12-19(29)35-17-16(28)25-...        NaN     NaN   

  Antibacterial  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN  

[5 rows x 26 columns]

In [ ]:
## 4. peptides.csv
* 

In [46]:
# ## Load the file, skipping the metadata line
# ## Read each individual csv files
# peptides_df = pd.read_csv(raw_files_path + 'peptides.csv', skiprows=1)

# peptides_df.head(5)

Ligand id                      Name     Species  \
0       5081              4-1BB ligand       Human   
1       5183  5-LOX activating protein       Human   
2       3633                      ACTH       Human   
3       1331                      ACTH  Mouse, Rat   
4       4857                 activin A       Human   

                                        Type Subunit ids    Subunit names  \
0  Endogenous peptide in human, mouse or rat         NaN              NaN   
1  Endogenous peptide in human, mouse or rat         NaN              NaN   
2  Endogenous peptide in human, mouse or rat         NaN              NaN   
3  Endogenous peptide in human, mouse or rat         NaN              NaN   
4  Endogenous peptide in human, mouse or rat        5006  inhibin &beta;A   

  Approved Withdrawn Labelled Radioactive  ...  Subunit Ensembl IDs  \
0      NaN       NaN      NaN         NaN  ...                  NaN   
1      NaN       NaN      NaN         NaN  ...                  NaN   
2      yes       NaN      NaN         NaN  ...                  NaN   
3      NaN       NaN      NaN         NaN  ...                  NaN   
4      NaN       NaN      NaN         NaN  ...      ENSG00000122641   

             INN                  Single letter amino acid sequence  \
0            NaN  MEYASDASLDPEAPWPPAPRARACRVLPWALVAGLLLLLLLAAACA...   
1            NaN  MDQETVGNVVLLAIVTLISVVQNGFFAHKVEHESRTQNGRSFQRTG...   
2  corticotropin            SYSMEHFRWGKPVGKKRRPVKVYPNGAEDESAEAFPLEF   
3            NaN            SYSMEHFRWGKPVGKKRRPVKVYPNVAENESAEAFPLEF   
4            NaN                                                NaN   

                    Three letter amino acid sequence HELM  \
0                                                NaN  NaN   
1                                                NaN  NaN   
2  Ser-Tyr-Ser-Met-Glu-His-Phe-Arg-Trp-Gly-Lys-Pr...  NaN   
3  Ser-Tyr-Ser-Met-Glu-His-Phe-Arg-Trp-Gly-Lys-Pr...  NaN   
4                                                NaN  NaN   

  Post-translational modification Chemical modification  \
0                             NaN                   NaN   
1                             NaN                   NaN   
2                           None.                   NaN   
3                           None.                   NaN   
4                             NaN                   NaN   

                                              SMILES  \
0                                                NaN   
1                                                NaN   
2  NCCCC[C@@H](C(=O)N[C@H](C(=O)N[C@H](C(=O)N1CCC...   
3                                                NaN   
4                                                NaN   

                      InChIKey  \
0                          NaN   
1                          NaN   
2  IDLFZVILOHSSID-OVLDLUHVSA-N   
3                          NaN   
4                          NaN   

                                               InChI  
0                                                NaN  
1                                                NaN  
2  InChI=1S/C207H308N56O58S/c1-108(2)89-140(186(3...  
3                                                NaN  
4                                                NaN  

[5 rows x 26 columns]

In [12]:
# list(peptides_df)

In [13]:
# ## check all unique combinations between subject & object
# from itertools import combinations

# # Get all combinations of 3 column names
# col_combos = ['Species']

# unique_combinations = peptides_df[col_combos].drop_duplicates()

# print(unique_combinations)

In [15]:
# ## Read each individual csv files
## use the detailed file instead
# df1 = pd.read_csv(raw_files_path + 'approved_drug_primary_target_interactions.csv')

# df1.head(10)

# GtoPdb Version: 2025.3 - published: 2025-09-10
Ligand        Ligand ID Target                     Target ID Target Ligand Target Ligand ID Target Species                                             Type
asenapine     22        5-HT<sub>2A</sub> receptor 6         NaN           0                Human                                         Synthetic organic
                        D<sub>2</sub> receptor     215       NaN           0                Human                                         Synthetic organic
apomorphine   33        D<sub>2</sub> receptor     215       NaN           0                Human                                         Synthetic organic
                        D<sub>3</sub> receptor     216       NaN           0                Human                                         Synthetic organic
                        D<sub>4</sub> receptor     217       NaN           0                Human                                         Synthetic organic
aripiprazole  34        D<sub>2</sub> receptor     215       NaN           0                Human                                         Synthetic organic
bromocriptine 35        D<sub>2</sub> receptor     215       NaN           0                Human                                         Synthetic organic
                        D<sub>3</sub> receptor     216       NaN           0                Human                                         Synthetic organic
                        D<sub>4</sub> receptor     217       NaN           0                Human                                         Synthetic organic

In [19]:
## Read each individual csv files
## Note: need to ignore the header which contains the GtoPdb version
edges_df = pd.read_csv(raw_files_path + 'approved_drug_detailed_interactions.csv', skiprows=1)

edges_df.head(10)

Ligand  Ligand ID               Type  \
0    asenapine         22  Synthetic organic   
1    asenapine         22  Synthetic organic   
2    asenapine         22  Synthetic organic   
3    asenapine         22  Synthetic organic   
4    asenapine         22  Synthetic organic   
5    asenapine         22  Synthetic organic   
6    asenapine         22  Synthetic organic   
7  apomorphine         33  Synthetic organic   
8  apomorphine         33  Synthetic organic   
9  apomorphine         33  Synthetic organic   

                                     Ligand Synonyms  \
0  Secuado&reg; (transdermal asenapine) | ORG5222...   
1  Secuado&reg; (transdermal asenapine) | ORG5222...   
2  Secuado&reg; (transdermal asenapine) | ORG5222...   
3  Secuado&reg; (transdermal asenapine) | ORG5222...   
4  Secuado&reg; (transdermal asenapine) | ORG5222...   
5  Secuado&reg; (transdermal asenapine) | ORG5222...   
6  Secuado&reg; (transdermal asenapine) | ORG5222...   
7  Kynmobi&reg; (sublingual film) | Apokyn&reg; |...   
8  Kynmobi&reg; (sublingual film) | Apokyn&reg; |...   
9  Kynmobi&reg; (sublingual film) | Apokyn&reg; |...   

                                         SMILES  CAS Number  \
0  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   
1  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   
2  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   
3  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   
4  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   
5  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   
6  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   
7          CN1CCc2c3[C@H]1Cc1ccc(c(c1c3ccc2)O)O     58-00-4   
8          CN1CCc2c3[C@H]1Cc1ccc(c(c1c3ccc2)O)O     58-00-4   
9          CN1CCc2c3[C@H]1Cc1ccc(c(c1c3ccc2)O)O     58-00-4   

                                Clinical Use Comment Bioactivity Comment  \
0  Asenapine is used to treat the symptoms of psy...                 NaN   
1  Asenapine is used to treat the symptoms of psy...                 NaN   
2  Asenapine is used to treat the symptoms of psy...                 NaN   
3  Asenapine is used to treat the symptoms of psy...                 NaN   
4  Asenapine is used to treat the symptoms of psy...                 NaN   
5  Asenapine is used to treat the symptoms of psy...                 NaN   
6  Asenapine is used to treat the symptoms of psy...                 NaN   
7  Used in the treatment of hypomobility associat...                 NaN   
8  Used in the treatment of hypomobility associat...                 NaN   
9  Used in the treatment of hypomobility associat...                 NaN   

                       Target  Target ID Target Entrez Gene ID  \
0  5-HT<sub>1A</sub> receptor          1                  3350   
1  5-HT<sub>1B</sub> receptor          2                  3351   
2  5-HT<sub>1D</sub> receptor          3                  3352   
3  5-ht<sub>1e</sub> receptor          4                  3354   
4  5-HT<sub>2A</sub> receptor          6                  3356   
5      D<sub>2</sub> receptor        215                  1813   
6      H<sub>1</sub> receptor        262                  3269   
7  5-HT<sub>1A</sub> receptor          1                  3350   
8  5-HT<sub>2A</sub> receptor          6                  3356   
9  5-HT<sub>2B</sub> receptor          7                  3357   

  Target Ensembl Gene ID Target Gene Name Target Ligand  Target Ligand ID  \
0        ENSG00000178394            HTR1A           NaN                 0   
1        ENSG00000135312            HTR1B           NaN                 0   
2        ENSG00000179546            HTR1D           NaN                 0   
3        ENSG00000168830            HTR1E           NaN                 0   
4        ENSG00000102468            HTR2A           NaN                 0   
5        ENSG00000149295             DRD2           NaN                 0   
6        ENSG00000196639             HRH1           NaN          

In [20]:
print(len(edges_df))
print(edges_df.columns)

4610
Index(['Ligand', 'Ligand ID', 'Type', 'Ligand Synonyms', 'SMILES',
       'CAS Number', 'Clinical Use Comment', 'Bioactivity Comment', 'Target',
       'Target ID', 'Target Entrez Gene ID', 'Target Ensembl Gene ID',
       'Target Gene Name', 'Target Ligand', 'Target Ligand ID',
       'Target Species'],
      dtype='object')


## Question 1:
* Can we use any of the specific types of medications?
* use the column "type" in class:drug to store such information, https://biolink.github.io/biolink-model/Drug/
* after discussion with Sierra and Vlado:
* short term, we will add those information under "gtopdb_ligand_type" column in the node file, it wont be used in translator UI for now
* long term, map the types to Chebi roles
* ```yaml
  * chemical role:
      is_a: attribute
      description: >-
        A role played by the molecular entity or part thereof within a chemical context.
      id_prefixes:
        - CHEBI
      exact_mappings:
        - CHEBI:51086
      examples:
        - value: CHEBI:35469  # antidepressant role

In [21]:
## Check the knowledge_source column again
## Count occurrences of each unique value in 'knowledge_source'
counts = edges_df['Type'].value_counts()

print(counts)

Type
Synthetic organic    3706
Peptide               248
Natural product       232
Antibody              213
Metabolite            181
Nucleic acid           21
Inorganic               9
Name: count, dtype: int64


## End of question 1

### Now start to process the edge df

In [24]:
## add prefix to both subject identifier and object identifier

## deep copy a df
edges_deep_copy_df = edges_df.copy(deep=True)

## Initialize new column with default values (original id as string)
edges_deep_copy_df['subject'] = edges_deep_copy_df['CAS Number'].astype(str)

## Add the CAS_number prefix to all values in this column
edges_deep_copy_df.loc[:, 'subject'] = "CAS:" + edges_deep_copy_df.loc[:, 'subject']

## Initialize new column with default values (original id as string)
edges_deep_copy_df['object'] = edges_deep_copy_df['Target Ensembl Gene ID'].astype(str)

## Add the CAS_number prefix to all values in this column
edges_deep_copy_df.loc[:, 'object'] = "ENSEMBL:" + edges_deep_copy_df.loc[:, 'object']

In [26]:
## Assign subject_category and object_category
## all subjects are medications, and objects are genes
edges_deep_copy_df['subject_category'] = 'biolink:Drug'
edges_deep_copy_df['object_category'] = 'biolink:Gene'

In [28]:
## add a relationship column to avoid conflicts with future sources
edges_deep_copy_df['relation'] = 'gtopdb_interactions'

In [29]:
edges_deep_copy_df.head(5)

Ligand  Ligand ID               Type  \
0  asenapine         22  Synthetic organic   
1  asenapine         22  Synthetic organic   
2  asenapine         22  Synthetic organic   
3  asenapine         22  Synthetic organic   
4  asenapine         22  Synthetic organic   

                                     Ligand Synonyms  \
0  Secuado&reg; (transdermal asenapine) | ORG5222...   
1  Secuado&reg; (transdermal asenapine) | ORG5222...   
2  Secuado&reg; (transdermal asenapine) | ORG5222...   
3  Secuado&reg; (transdermal asenapine) | ORG5222...   
4  Secuado&reg; (transdermal asenapine) | ORG5222...   

                                         SMILES  CAS Number  \
0  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   
1  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   
2  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   
3  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   
4  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   

                                Clinical Use Comment Bioactivity Comment  \
0  Asenapine is used to treat the symptoms of psy...                 NaN   
1  Asenapine is used to treat the symptoms of psy...                 NaN   
2  Asenapine is used to treat the symptoms of psy...                 NaN   
3  Asenapine is used to treat the symptoms of psy...                 NaN   
4  Asenapine is used to treat the symptoms of psy...                 NaN   

                       Target  Target ID  ... Target Ensembl Gene ID  \
0  5-HT<sub>1A</sub> receptor          1  ...        ENSG00000178394   
1  5-HT<sub>1B</sub> receptor          2  ...        ENSG00000135312   
2  5-HT<sub>1D</sub> receptor          3  ...        ENSG00000179546   
3  5-ht<sub>1e</sub> receptor          4  ...        ENSG00000168830   
4  5-HT<sub>2A</sub> receptor          6  ...        ENSG00000102468   

  Target Gene Name Target Ligand Target Ligand ID  Target Species  \
0            HTR1A           NaN                0           Human   
1            HTR1B           NaN                0           Human   
2            HTR1D           NaN                0           Human   
3            HTR1E           NaN                0           Human   
4            HTR2A           NaN                0           Human   

          subject                   object subject_category object_category  \
0  CAS:65576-45-6  ENSEMBL:ENSG00000178394     biolink:Drug    biolink:Gene   
1  CAS:65576-45-6  ENSEMBL:ENSG00000135312     biolink:Drug    biolink:Gene   
2  CAS:65576-45-6  ENSEMBL:ENSG00000179546     biolink:Drug    biolink:Gene   
3  CAS:65576-45-6  ENSEMBL:ENSG00000168830     biolink:Drug    biolink:Gene   
4  CAS:65576-45-6  ENSEMBL:ENSG00000102468     biolink:Drug    biolink:Gene   

              relation  
0  gtopdb_interactions  
1  gtopdb_interactions  
2  gtopdb_interactions  
3  gtopdb_interactions  
4  gtopdb_interactions  

[5 rows x 21 columns]

### Load the Biolink category and predicate dictionary for mapping subject, object, and predicate types

In [30]:
## Load the Biolink category and predicate dictionary for mapping subject, object, and predicate types
%run ./Biolink_category_and_predication_dictionary.ipynb

Date of last update:  2025-09-04
Order is to always process Node/category map first, since the Edeg/predicate map depends on biolink-complainat node values
-----------------------------------------------------------------------------------------------------------------------------
Dictionary: category_map, Key template: Subject_category or Object_category
------------------------------------------------------------------------------------------
Dictionary: predicate_map, Key template: (Subject_category, Object_category, Predicate)


In [63]:
## deep copy a df
edges_deep_copy_df2 = edges_deep_copy_df.copy(deep=True)

In [64]:
## start convert to biolink compliant predicate types based on display_relation column
## check supplementary table of primeKG, descriptions on relations
## translator discussion point - data ingestion group meeting
## separate out off label drug usage, clinical trial, and FDA approved usage
## for 'synergistic interaction', combine with knowledge from LLM clinical trial attempts and find drug combination and which disease 
## they targeted
## currently exclude associated with
## not need to include 'phenotype absent'

## match only combinations allowed in the pair dictionary
## Apply the mapping and return a Series with 3 columns
edges_deep_copy_df2[['predicate', 'subject_direction_qualifier', 'object_direction_qualifier', 'object_aspect_qualifier']] = edges_deep_copy_df2.apply(
    lambda row: pd.Series(
        predicate_map.get(
            (row['subject_category'], row['object_category'], row['relation']),
            [None, None, None]  # Default if not found
        )
    ),
    axis=1
)

In [65]:
edges_deep_copy_df2.head(5)

Ligand  Ligand ID               Type  \
0  asenapine         22  Synthetic organic   
1  asenapine         22  Synthetic organic   
2  asenapine         22  Synthetic organic   
3  asenapine         22  Synthetic organic   
4  asenapine         22  Synthetic organic   

                                     Ligand Synonyms  \
0  Secuado&reg; (transdermal asenapine) | ORG5222...   
1  Secuado&reg; (transdermal asenapine) | ORG5222...   
2  Secuado&reg; (transdermal asenapine) | ORG5222...   
3  Secuado&reg; (transdermal asenapine) | ORG5222...   
4  Secuado&reg; (transdermal asenapine) | ORG5222...   

                                         SMILES  CAS Number  \
0  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   
1  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   
2  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   
3  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   
4  Clc1ccc2c(c1)[C@H]1CN(C[C@@H]1c1c(O2)cccc1)C  65576-45-6   

                                Clinical Use Comment Bioactivity Comment  \
0  Asenapine is used to treat the symptoms of psy...                 NaN   
1  Asenapine is used to treat the symptoms of psy...                 NaN   
2  Asenapine is used to treat the symptoms of psy...                 NaN   
3  Asenapine is used to treat the symptoms of psy...                 NaN   
4  Asenapine is used to treat the symptoms of psy...                 NaN   

                       Target  Target ID  ... Target Species         subject  \
0  5-HT<sub>1A</sub> receptor          1  ...          Human  CAS:65576-45-6   
1  5-HT<sub>1B</sub> receptor          2  ...          Human  CAS:65576-45-6   
2  5-HT<sub>1D</sub> receptor          3  ...          Human  CAS:65576-45-6   
3  5-ht<sub>1e</sub> receptor          4  ...          Human  CAS:65576-45-6   
4  5-HT<sub>2A</sub> receptor          6  ...          Human  CAS:65576-45-6   

                    object subject_category  object_category  \
0  ENSEMBL:ENSG00000178394     biolink:Drug     biolink:Gene   
1  ENSEMBL:ENSG00000135312     biolink:Drug     biolink:Gene   
2  ENSEMBL:ENSG00000179546     biolink:Drug     biolink:Gene   
3  ENSEMBL:ENSG00000168830     biolink:Drug     biolink:Gene   
4  ENSEMBL:ENSG00000102468     biolink:Drug     biolink:Gene   

              relation                      predicate  \
0  gtopdb_interactions  biolink:DrugToGeneAssociation   
1  gtopdb_interactions  biolink:DrugToGeneAssociation   
2  gtopdb_interactions  biolink:DrugToGeneAssociation   
3  gtopdb_interactions  biolink:DrugToGeneAssociation   
4  gtopdb_interactions  biolink:DrugToGeneAssociation   

  subject_direction_qualifier object_direction_qualifier  \
0                        None                       None   
1                        None                       None   
2                        None                       None   
3                        None                       None   
4                        None                       None   

  object_aspect_qualifier  
0                    None  
1                    None  
2                    None  
3                    None  
4                    None  

[5 rows x 25 columns]

In [66]:
## add a new knowledge_souce column and set value to be "PrimeKG"
edges_deep_copy_df2['knowledge_source'] = 'GtoPdB'
## add a new knowledge_level column and set value to be 'knowledge_assertion'
edges_deep_copy_df2['knowledge_level'] = 'knowledge_assertion'
## add a new agent_type column and set value to be 'manual_agent'
edges_deep_copy_df2['agent_type'] = 'automated_agent'

## copy to a final df
edge_df = edges_deep_copy_df2.copy(deep = True)

print(edge_df.shape[0])
## Remove rows where subject or object is empty
# Remove rows where 'Subject' OR 'Object' have NaN values
edge_df = edge_df.dropna(subset=['subject', 'object'])

print(edge_df.shape[0])

edge_df['deploy_date'] = deployment_date
# edge_df.head(5)

4610
4610


In [67]:
## create a publication column and fill na
## if all of them are empty then fill na
edge_df['publications'] = np.nan

In [68]:
## rename those columns into desired format
edge_df.rename(columns={'Ligand': 'subject_name', 'Target Gene Name': 'object_name'}, inplace=True)

In [69]:
## create a context_qualifier column and fill na
## if all of them are empty then fill na
edge_df['context_qualifier'] = np.nan
# edge_df.head(5)

In [70]:
### Add resources_id column, checking whether edge is already
column_list = ['subject', 'predicate', 'object', 'context_qualifier', 'deploy_date']
# Apply the function to each row to generate UUIDs
edge_df['id'] = edge_df[column_list].apply(generate_uuid, axis=1)

# edge_df['id'] = generate_uuid_from_columns(edge_df, column_list)
# edge_df.head(5)

### Now create the corresponding node file
* only need three columns: id, name, category

In [71]:
node_subject_df = edge_df[['subject', 'subject_name', 'subject_category', 'Type']]
node_object_df = edge_df[['object', 'object_name', 'object_category', 'Target Species']]

In [72]:
## rename those columns into desired format
node_subject_df.rename(columns={'Type': 'gtopdb_ligand_type'}, inplace=True)

## rename those columns into desired format
node_object_df.rename(columns={'Target Species': 'gtopdb_species_type'}, inplace=True)

/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_21116/3136485221.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  node_subject_df.rename(columns={'Type': 'gtopdb_ligand_type'}, inplace=True)
/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_21116/3136485221.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  node_object_df.rename(columns={'Target Species': 'gtopdb_species_type'}, inplace=True)


In [73]:
## rename those columns into desired format
node_subject_df.rename(columns={'subject': 'id', 'subject_name': 'name', 'subject_category': 'category'}, inplace=True)
node_object_df.rename(columns={'object': 'id', 'object_name': 'name', 'object_category': 'category'}, inplace=True)

concat_node_df = pd.concat([node_subject_df, node_object_df]).drop_duplicates(keep='first')

concat_node_df.head()

/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_21116/675656324.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  node_subject_df.rename(columns={'subject': 'id', 'subject_name': 'name', 'subject_category': 'category'}, inplace=True)
/var/folders/ml/cjwyk2ps62361rr96_3b_n080000gp/T/ipykernel_21116/675656324.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  node_object_df.rename(columns={'object': 'id', 'object_name': 'name', 'object_category': 'category'}, inplace=True)


id           name      category gtopdb_ligand_type  \
0    CAS:65576-45-6      asenapine  biolink:Drug  Synthetic organic   
7       CAS:58-00-4    apomorphine  biolink:Drug  Synthetic organic   
22  CAS:129722-12-9   aripiprazole  biolink:Drug  Synthetic organic   
34   CAS:25614-03-3  bromocriptine  biolink:Drug  Synthetic organic   
53   CAS:36505-84-7      buspirone  biolink:Drug  Synthetic organic   

   gtopdb_species_type  
0                  NaN  
7                  NaN  
22                 NaN  
34                 NaN  
53                 NaN

In [74]:
print(edge_df.columns)

Index(['subject_name', 'Ligand ID', 'Type', 'Ligand Synonyms', 'SMILES',
       'CAS Number', 'Clinical Use Comment', 'Bioactivity Comment', 'Target',
       'Target ID', 'Target Entrez Gene ID', 'Target Ensembl Gene ID',
       'object_name', 'Target Ligand', 'Target Ligand ID', 'Target Species',
       'subject', 'object', 'subject_category', 'object_category', 'relation',
       'predicate', 'subject_direction_qualifier',
       'object_direction_qualifier', 'object_aspect_qualifier',
       'knowledge_source', 'knowledge_level', 'agent_type', 'deploy_date',
       'publications', 'context_qualifier', 'id'],
      dtype='object')


In [75]:
## drop no longer needed columns from edge df
print(edge_df.columns.tolist())
drop_cols = ['Ligand ID', 'Type', 'Ligand Synonyms', 'SMILES',
       'CAS Number', 'Clinical Use Comment', 'Bioactivity Comment', 'Target',
       'Target ID', 'Target Entrez Gene ID', 'Target Ensembl Gene ID',
       'Target Ligand', 'Target Ligand ID', 'Target Species',
       'relation',]

edge_output_df = edge_df.drop(drop_cols, axis=1)

edge_output_df.head()

['subject_name', 'Ligand ID', 'Type', 'Ligand Synonyms', 'SMILES', 'CAS Number', 'Clinical Use Comment', 'Bioactivity Comment', 'Target', 'Target ID', 'Target Entrez Gene ID', 'Target Ensembl Gene ID', 'object_name', 'Target Ligand', 'Target Ligand ID', 'Target Species', 'subject', 'object', 'subject_category', 'object_category', 'relation', 'predicate', 'subject_direction_qualifier', 'object_direction_qualifier', 'object_aspect_qualifier', 'knowledge_source', 'knowledge_level', 'agent_type', 'deploy_date', 'publications', 'context_qualifier', 'id']


subject_name object_name         subject                   object  \
0    asenapine       HTR1A  CAS:65576-45-6  ENSEMBL:ENSG00000178394   
1    asenapine       HTR1B  CAS:65576-45-6  ENSEMBL:ENSG00000135312   
2    asenapine       HTR1D  CAS:65576-45-6  ENSEMBL:ENSG00000179546   
3    asenapine       HTR1E  CAS:65576-45-6  ENSEMBL:ENSG00000168830   
4    asenapine       HTR2A  CAS:65576-45-6  ENSEMBL:ENSG00000102468   

  subject_category object_category                      predicate  \
0     biolink:Drug    biolink:Gene  biolink:DrugToGeneAssociation   
1     biolink:Drug    biolink:Gene  biolink:DrugToGeneAssociation   
2     biolink:Drug    biolink:Gene  biolink:DrugToGeneAssociation   
3     biolink:Drug    biolink:Gene  biolink:DrugToGeneAssociation   
4     biolink:Drug    biolink:Gene  biolink:DrugToGeneAssociation   

  subject_direction_qualifier object_direction_qualifier  \
0                        None                       None   
1                        None                       None   
2                        None                       None   
3                        None                       None   
4                        None                       None   

  object_aspect_qualifier knowledge_source      knowledge_level  \
0                    None           GtoPdB  knowledge_assertion   
1                    None           GtoPdB  knowledge_assertion   
2                    None           GtoPdB  knowledge_assertion   
3                    None           GtoPdB  knowledge_assertion   
4                    None           GtoPdB  knowledge_assertion   

        agent_type deploy_date  publications  context_qualifier  \
0  automated_agent  2026-01-06           NaN                NaN   
1  automated_agent  2026-01-06           NaN                NaN   
2  automated_agent  2026-01-06           NaN                NaN   
3  automated_agent  2026-01-06           NaN                NaN   
4  automated_agent  2026-01-06           NaN                NaN   

                                     id  
0  7e796393-0df8-50d7-ae7a-fccce6478f8e  
1  308e522c-c5d6-5bd7-bd51-0dce57ca2b62  
2  44e7ecea-962b-5336-8809-ba770ee36656  
3  7bfe056f-2f5f-5266-b97e-6f51b5e388db  
4  c6dcf08d-3ce9-5c77-aeb3-06c2d3ccd860

In [76]:
## exclude rows with Null and empty string values in the predicate column
# edge_output_df = edge_output_df.dropna(subset=['predicate'])

## Drop rows where 'name' is NaN, None, or empty string
edge_output_df = edge_output_df[~edge_output_df['predicate'].isna() & (edge_output_df['predicate'].str.strip() != '')]

In [77]:
## Check the knowledge_source column again
## Count occurrences of each unique value in 'knowledge_source'
counts = edge_output_df['knowledge_source'].value_counts()

print(counts)

knowledge_source
GtoPdB    4610
Name: count, dtype: int64


In [78]:
## reorder the dataframe
desired_order = [
    'subject', 'subject_name', 'subject_category', 'object', 'object_name', 'object_category', 
    'predicate', 'subject_direction_qualifier', 'object_direction_qualifier', 'object_aspect_qualifier',
    'knowledge_source', 'publications', 'knowledge_level', 'context_qualifier', 'agent_type', 
    'deploy_date', 'id'
]

edge_output_df = edge_output_df[desired_order]

print(edge_output_df.columns.tolist())

['subject', 'subject_name', 'subject_category', 'object', 'object_name', 'object_category', 'predicate', 'subject_direction_qualifier', 'object_direction_qualifier', 'object_aspect_qualifier', 'knowledge_source', 'publications', 'knowledge_level', 'context_qualifier', 'agent_type', 'deploy_date', 'id']


### Now quality control of the parsed PrimeKG dataframe
* Original edge counts: 4610
* QA:
  - dropped duplicated edges in reverse direction
  - dropped enzyme and node types not easily mapped in biolink
  - removed edges with invalid predicate, subject, and object relationship

In [79]:
## Create a graph from the DataFrame
graph = nx.from_pandas_edgelist(edge_output_df, 'subject', 'object', edge_attr='predicate')

## Print graph information
print('Number of nodes', len(set(graph.nodes)))
print('Number of edges', len(set(graph.edges)))
print('Average degree', sum(dict(graph.degree).values()) / len(graph.nodes))

Number of nodes 2705
Number of edges 4194
Average degree 3.100924214417745


In [80]:
## check all unique predicate values
counts = concat_node_df['category'].value_counts()
print(counts)

category
biolink:Drug    2139
biolink:Gene     959
Name: count, dtype: int64


In [81]:
## check missingness
print(edge_df[['predicate', 'subject_direction_qualifier', 'object_direction_qualifier', 'object_aspect_qualifier']].isna().sum())

predicate                         0
subject_direction_qualifier    4610
object_direction_qualifier     4610
object_aspect_qualifier        4610
dtype: int64


In [82]:
## fill na temporily to get the combo number
filled = edge_df[['predicate', 'subject_direction_qualifier', 'object_direction_qualifier', 'object_aspect_qualifier']].fillna('None')
counts = filled.value_counts().reset_index(name='count')
print(counts)

                       predicate subject_direction_qualifier  \
0  biolink:DrugToGeneAssociation                        None   

  object_direction_qualifier object_aspect_qualifier  count  
0                       None                    None   4610  


## Now download the concatenated node & edge files

In [84]:
## Define the output path for node & edge files after formatting
download_path_node_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/wQualityControl/GtoPdB_approved_drug_detailed_interactions_parsed_node_{version_number}.tsv'
download_path_edge_file = f'/Users/Weiqi0/ISB_working/Ilya_lab/Translator/Pharmagenomics_KG/files/parsed/wQualityControl/GtoPdB_approved_drug_detailed_interactions_parsed_edge_{version_number}.tsv'

## download both node and edge files
## Download the result df
concat_node_df.to_csv(download_path_node_file, sep ='\t', index=False)
edge_output_df.to_csv(download_path_edge_file, sep ='\t', index=False)